src/hip_memory.cpp

```diff
@@ -341,6 +345,14 @@ void memcpy_impl(void* __restrict dst, const void* __restrict src, size_t n,
     case hipMemcpyHostToDevice: return h2d_copy(dst, src, n, di);
     case hipMemcpyDeviceToHost: return d2h_copy(dst, src, n, si);
     case hipMemcpyDeviceToDevice: {
+        hsa_status_t res = hsa_amd_agents_allow_access(1u, &si.agentOwner,
```

Comment:
This is not writtent well. Assume that the `hsa_amd_agents_allow_access` call succeeds. Then you're going to call it again immediately because... reasons. Try to write this in affirmative form, it'll prevent such issues i.e. if allow access succeeds return do_copy, and subsequently handle the special case. This is also a matter of logical hygiene (follow the likeliest path first).